onnx/optimizer/passes/fuse_bn_into_conv.h

```diff
@@ -0,0 +1,174 @@
+        for (int64_t i = 0; i < eps.sizes()[0]; i++)  {
+          bc.floats().push_back(0.f);
+        }
+
+        if (conv_inputs.size() == 3) {
+          auto bc_iter = graph.getInitializer(conv_inputs[2]->uniqueName());
+          if (bc_iter == end_iter) {
+            return false;
+          }
+          bc = *bc_iter;
+          ONNX_ASSERT(bc.sizes().size() == 1 && bc.sizes()[0] == s.sizes()[0]);
+        }
+
+        var.add(eps);
+        var.sqrt();
+        s.divide(var);
+        if (!s.is_raw_data()) {
+          const char * ptr = reinterpret_cast<const char *>(s.floats().data());
+          std::string string_rep(ptr, ptr + sizeof(float) * s.floats().size());
+          s.set_raw_data(string_rep);
+        }
+
+        W.scale_by_first_dim(s);
+        bc.subtract(m);
+        bc.multiply(s);
+        bc.add(bbn);
+        break;
+      }
+      case ONNX_NAMESPACE::TensorProto_DataType_DOUBLE:
+      {
+        eps.sizes().push_back(s.sizes()[0]);
+        eps.elem_type() = ONNX_NAMESPACE::TensorProto_DataType_DOUBLE;
+        for (int64_t i = 0; i < eps.sizes()[0]; i++)  {
+          eps.doubles().push_back((double)epsilon);
+        }
+        bc.sizes().push_back(s.sizes()[0]);
+        bc.elem_type() = ONNX_NAMESPACE::TensorProto_DataType_DOUBLE;
+        for (int64_t i = 0; i < eps.sizes()[0]; i++)  {
+          bc.doubles().push_back(0.);
+        }
+
+        if (conv_inputs.size() == 3) {
+          auto bc_iter = graph.getInitializer(conv_inputs[2]->uniqueName());
+          if (bc_iter == end_iter) {
+            return false;
+          }
+          bc = *bc_iter;
+          ONNX_ASSERT(bc.sizes().size() == 1 && bc.sizes()[0] == s.sizes()[0]);
+        }
+
+        var.add(eps);
+        var.sqrt();
+        s.divide(var);
+        if (!s.is_raw_data()) {
+          const char * ptr = reinterpret_cast<const char *>(s.doubles().data());
+          std::string string_rep(ptr, ptr + sizeof(double) * s.doubles().size());
+          s.set_raw_data(string_rep);
+        }
+        W.scale_by_first_dim(s);
+        bc.subtract(m);
+        bc.multiply(s);
+        bc.add(bbn);
+        break;
+      }
+      default:
+        throw("Incompatible data type");
+    }
+
+    replace_inputs(W, bc, conv, graph);
+    return true;
+  }
+
+  void fuse_bn_into_conv(Graph& graph) {
+
+    for (auto it = graph.begin(); it != graph.end(); ++it) {
+      auto* n = *it;
+      DescendOnGraphAttributes(n, [this](Graph& g){fuse_bn_into_conv(g);});
+      if (n->kind() == kBatchNormalization && n->inputs()[0]->node()->kind() == kConv) {
+        auto origInput = n->inputs()[0];
+        if (origInput->uses().size() > 1 ||
+            n->outputs().size() > 1 ||
+            !modify_conv(origInput->node(), n, graph)) {
+          continue;
+        }
+        for (int i = 1; i <= 4; i++)  {
+          if (n->inputs()[i]->uses().size() == 0) {
+            graph.eraseInitializerAndInput(n->inputs()[i]);
```

Comment:
You need to check whether these inputs are really in initializers.

Reply:
this is checked in modify_conv - if not initializers then continue; so this function is not run